include/vara/Feature/FeatureModel.h

```diff
@@ -68,29 +69,135 @@ class FeatureModel {
     }
   };
 
-  FeatureModelIter begin() { return FeatureModelIter(Features.begin()); }
+public:
+  using feature_model_iterator = FeatureModelIter;
+  using const_feature_model_iterator = FeatureModelIter;
+
+  feature_model_iterator begin() { return FeatureModelIter(Features.begin()); }
+
+  [[nodiscard]] const_feature_model_iterator begin() const {
+    return FeatureModelIter(Features.begin());
+  }
 
   FeatureModelIter end() { return FeatureModelIter(Features.end()); }
 
+  [[nodiscard]] const_feature_model_iterator end() const {
+    return FeatureModelIter(Features.end());
+  }
+
+  [[nodiscard]] unsigned int size() { return Features.size(); }
+
+  void view() { ViewGraph(this, "FeatureModel-" + this->getName()); }
+
+  Feature *getFeature(llvm::StringRef Name) { return Features[Name].get(); }
+
   LLVM_DUMP_METHOD
-  void dump() const {
-    for (const auto &F : this->Features) {
-      F.second->dump();
-      llvm::outs() << "\n";
+  void dump() const;
+
+  class FeatureModelBuilder {
+    using EdgeMapType =
+        typename llvm::StringMap<llvm::SmallVector<std::string, 3>>;
+
+    std::string VmName;
+    fs::path Path;
+    Feature *Root{nullptr};
+    FeatureMapTy Features;
+    FeatureModel::ConstraintsTy Constraints;
+    llvm::StringMap<std::string> Parents;
+    EdgeMapType Children;
+    EdgeMapType Excludes;
+    EdgeMapType Implications;
+    EdgeMapType Alternatives;
+
+  public:
+    FeatureModelBuilder() = default;
+
+    void init() {
+      VmName = "";
+      Path = "";
+      Root = nullptr;
+      Features.clear();
+      Constraints.clear();
+      Parents.clear();
+      Children.clear();
+      Excludes.clear();
+      Implications.clear();
+      Alternatives.clear();
     }
-    for (auto C : this->Constraints) {
-      for (auto F : C) {
-        if (!F.second) {
-          llvm::outs() << "!";
-        }
-        llvm::outs() << F.first->getName();
-        if (F != C.back()) {
-          llvm::outs() << " | ";
-        }
-      }
-      llvm::outs() << "\n";
+
+    bool addFeature(const std::string &Key,
+                    const NumericFeature::ValuesVariantType &Values,
+                    bool Opt = false,
+                    std::optional<FeatureSourceRange> Loc = std::nullopt) {
+      return Features
+          .try_emplace(Key, std::make_unique<NumericFeature>(Key, Values, Opt,
+                                                             std::move(Loc)))
+          .second;
     }
-  }
+
+    bool addFeature(const std::string &Key, bool Opt = false,
+                    std::optional<FeatureSourceRange> Loc = std::nullopt) {
+      return Features
+          .try_emplace(
+              Key, std::make_unique<BinaryFeature>(Key, Opt, std::move(Loc)))
+          .second;
+    }
+
+    FeatureModelBuilder *addChild(const std::string &P, const std::string &C) {
+      Children[P].push_back(C);
+      Parents[C] = P;
+      return this;
+    }
```

Comment:
Does it make sense to also add overloads for when we have a `Feature &` and what to add a child etc.?

Reply:
This method / builder is supposed to be only used while parsing / iteratively building a FM, so it makes not that much sense to build those connections for an already existing `Feature` by hand or even hold such an object in the first place. I added `addFeature(Feature&)` instead which basically behaves like inserting all related connections one after another, which can be used to duplicate existing subtrees between different FMs.